tests/beman/task/promise_base.test.cpp

```diff
@@ -29,29 +36,64 @@ struct int_receiver {
     int& value;
     void set_value(int v) && noexcept { this->value = v; }
     void set_error(auto&&) && noexcept { unreachable("unexpected call to set_error"); }
+    void set_stopped() && noexcept { unreachable("unexpected call to set_stopped"); }
 };
 static_assert(::beman::execution::receiver<int_receiver>);
 } // namespace
 
+template <typename... E>
+struct env {
+    using error_types = ex::completion_signatures<ex::set_error_t(E)...>;
+};
+
+template <typename T, typename... E>
+struct state : bt::state_base<T, env<E...>> {
+    using Environment      = env<E...>;
+    using stop_source_type = ::beman::task::detail::stop_source_of_t<Environment>;
+    using stop_token_type  = decltype(std::declval<stop_source_type>().get_token());
+    using scheduler_type   = ::beman::task::detail::scheduler_of_t<Environment>;
+
+    stop_source_type source;
+    Environment      env;
+    bool             completed{};
+    bool             token{};
+    bool             got_environment{};
+
+    ::std::coroutine_handle<> do_complete() override {
+        this->completed = true;
+        return std::noop_coroutine();
+    }
+    stop_token_type do_get_stop_token() override {
+        this->token = true;
+        return this->source.get_token();
+    }
+    Environment& do_get_environment() override {
+        this->got_environment = true;
+        return this->env;
+    }
+    auto do_get_scheduler() -> scheduler_type override { return scheduler_type(bt::inline_scheduler()); }
+    auto do_set_scheduler(scheduler_type other) -> scheduler_type override {
+        return scheduler_type(bt::inline_scheduler());
+    }
+};
+
 int main() {
     {
-        beman::task::detail::promise_base<beman::task::detail::stoppable::no,
-                                          void,
-                                          beman::execution::completion_signatures<beman::execution::set_error_t(int)>>
-            result{};
+        state<void, int>                                                                      s{};
```

Comment:
that result type is a mouthful lol

```suggestion
        using result_type = beman::task::detail::promise_base<beman::task::detail::stoppable::no, void, env<int>>;
        state<void, int>                                                                      s{};
```